### PR TITLE
AB#2458 AWS NitroTPM attestation

### PR DESCRIPTION
**Pull request by leongross**
### Proposed change(s)

- Add NitroTPM attestation

### Additional info
- NitroTPM does not provide an `EKCert` for verification yet

### Checklist

- [ ] Update [CHANGELOG.md](https://github.com/edgelesssys/constellation/blob/main/CHANGELOG.md)
- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [ ] Link to Milestone

### CODE REVIEW

**Review by daniel-weisse — COMMENTED** [review]
No in depth review, just some quick hints

File: internal/attestation/aws/validator_test.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	// Since Amazon does not offer Linux stock AMI's with TPM support enabled, we use Windows images in this test.
+	// https://docs.aws.amazon.com/AWSEC2/latest/WindowsGuide/enable-nitrotpm-prerequisites.html
+	mockDocTpm := mockDocBase
+	mockDocTpm.ImageID = "ami-0e6b1588acc5a055f"
+	mockDocTpm.Region = "us-east-2"
+
+	// Stock aws linux (per default no TPM enabled), here could be the ami of ANY stock linux.
+	mockDocNoTpm := mockDocBase
+	mockDocNoTpm.ImageID = "ami-0b59bfac6be064b78"
+	mockDocTpm.Region = "us-east-2"
```
Comment (daniel-weisse) [inline internal/attestation/aws/validator_test.go:87]: Unit test should not make real API calls.
If you decide to implement the suggestion of rewriting the `tpmEnabled` function as a method of `Validator`, you can instead run the function against a stub AWS API.

File: internal/attestation/gcp/issuer.go [hunk 2]
```diff
@@ -26,11 +26,12 @@ type Issuer struct {
 
 // NewIssuer initializes a new GCP Issuer.
 func NewIssuer() *Issuer {
+	GetInstanceInfo := getGCEInstanceInfo(metadataClient{})
 	return &Issuer{
 		Issuer: vtpm.NewIssuer(
 			vtpm.OpenVTPM,
 			tpmclient.GceAttestationKeyRSA,
-			getGCEInstanceInfo(metadataClient{}),
+			GetInstanceInfo,
```
Comment (daniel-weisse) [inline internal/attestation/gcp/issuer.go:34]: Why this change?

File: internal/attestation/vtpm/attestation.go [hunk 3]
```diff
@@ -38,6 +38,11 @@ var (
 	// QEMUPCRSelection are the PCR values verified for QEMU based Contellations.
 	// PCR[1] is excluded. See: https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClient_PFP_r1p05_v23_pub.pdf#%5B%7B%22num%22:157,%22gen%22:0%7D,%7B%22name%22:%22XYZ%22%7D,33,400,0%5D
 	// PCR[10] is excluded since its value is derived from a digest of PCR[0-7]. See: https://sourceforge.net/p/linux-ima/wiki/Home/#ima-measurement-list
+	AWSPCRSelection = tpm2.PCRSelection{
```
Comment (daniel-weisse) [inline internal/attestation/vtpm/attestation.go:41]: Doc comment belongs to `QEMUPCRSelection`.

File: internal/attestation/vtpm/attestation.go [hunk 4]
```diff
@@ -38,6 +38,11 @@ var (
...
+		Hash: tpm2.AlgSHA256,
+		PCRs: []int{0, 2, 3, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23},
```
Comment (daniel-weisse) [inline internal/attestation/vtpm/attestation.go:43]: Did you check if these PCRs are stable for different instances and across reboots?
We would need to check using a Constellation image, which should be available for AWS by now.

**Review by daniel-weisse — CHANGES_REQUESTED** [review]

File: internal/attestation/aws/validator.go [hunk 5]
```diff
@@ -7,13 +7,85 @@ SPDX-License-Identifier: AGPL-3.0-only
...
+
+	// retrieve instanceIdentityDocument from attestation document
+	idDocument := imds.InstanceIdentityDocument{}
+	err := json.Unmarshal(attestation.UserData, &idDocument)
+	if err != nil {
+		return err
+	}
+
+	return tpmEnabled(idDocument)
 }
```
[the remark this answers is not in the document]
Comment (leongross) [reply internal/attestation/aws/validator.go:91]: As far as I see it we cannot simply drop-in replace this function, because we have to comply with the function definition of [`vtpm.ValidateCVM`](https://github.com/edgelesssys/constellation/blob/main/internal/attestation/vtpm/attestation.go#L61), which is parameterized differently. That's why I kept `validateCVM` as a wrapper function.

[the remark this answers is not in the document]
Comment (daniel-weisse) [reply internal/attestation/aws/validator.go:91]: If tpmEnabled is a method of Validator you can fulfill the function signature requirements as follows:

```Go

func (v *Validator) tpmEnabled(attestation vtpm.AttestationDocument) error {
	idDocument := imds.InstanceIdentityDocument{}
	err := json.Unmarshal(attestation.UserData, &idDocument)
	if err != nil {
		return err
	}
	// The rest of the function implementation
	// ...
}

File: internal/attestation/vtpm/attestation.go (hunk 4 above)
Comment (leongross) [reply internal/attestation/vtpm/attestation.go:43]: I tested it on stock Ubuntu 22.04 instances with `TpmSupport=v2.0`, but not yet with the constellation images, I will verify this .

File: internal/attestation/gcp/issuer.go (hunk 2 above)
Comment (daniel-weisse) [reply internal/attestation/gcp/issuer.go:34]: IMO instantiating the function implies you plan on using it more than once, which is not the case here.
Additionally private variables and functions should start with a lowercase character.

File: internal/attestation/aws/issuer.go [hunk 6]
```diff
@@ -21,14 +21,14 @@ import (
...
+// Issuer for AWS TPM attestation
 type Issuer struct {
 	oid.AWS
 	*vtpm.Issuer
 }
 
 func NewIssuer() *Issuer {
-	awsIMDS := imds.New(imds.Options{})
-	GetInstanceInfo := getInstanceInfo(awsIMDS)
+	GetInstanceInfo := getInstanceInfo(imds.New(imds.Options{}))
```
Comment (daniel-weisse) [inline internal/attestation/aws/issuer.go:31]:
```suggestion
	getInstanceInfo := getInstanceInfo(imds.New(imds.Options{}))
```
As already mentioned, I would prefer to not instantiate the function, but use it directly.
Regardless of final implementation, the variable should start with a lowercase character

File: internal/attestation/aws/issuer_test.go [hunk 7]
```diff
@@ -92,6 +104,12 @@ type fakeMetadataClient struct {
 }
 
 func (c *fakeMetadataClient) GetInstanceIdentityDocument(context.Context, *imds.GetInstanceIdentityDocumentInput, ...func(*imds.Options)) (*imds.GetInstanceIdentityDocumentOutput, error) {
+	testRegions := []string{
+		"us-east-2", "us-east-1", "us-west-1", "us-west-2",
+	}
+
+	fmt.Printf("%v", testRegions)
```
Comment (daniel-weisse) [inline internal/attestation/aws/issuer_test.go:111]:
```suggestion
```
Leftover debug code?

File: internal/attestation/aws/validator.go [hunk 8]
```diff
@@ -18,8 +18,10 @@ import (
 	"github.com/aws/aws-sdk-go-v2/service/ec2"
 	"github.com/edgelesssys/constellation/v2/internal/attestation/vtpm"
 	"github.com/edgelesssys/constellation/v2/internal/oid"
+	"github.com/google/go-tpm/tpm2"
 )
 
+// Issuer for AWS TPM attestation
```
Comment (daniel-weisse) [inline internal/attestation/aws/validator.go:24]:
```suggestion
// Validator for AWS TPM attestation.
```

**Review by daniel-weisse — COMMENTED** [review]
CI won't run as long as there are merge conflicts with the main branch.
The CI should automatically detect and annotate any linting issues, e.g. doc comments.

File: internal/attestation/gcp/issuer.go (hunk 2 above)
Comment (leongross) [reply internal/attestation/gcp/issuer.go:34]: I agree, it might imply wrong usage, I will change it back as it was.

File: internal/attestation/aws/issuer_test.go [hunk 9]
```diff
@@ -94,40 +103,17 @@ func TestGetInstanceInfo(t *testing.T) {
...
 }
 
-type fakeMetadataClient struct {
-	projectIDString    string
-	instanceNameString string
-	zoneString         string
-	projecIDErr        error
-	instanceNameErr    error
-	zoneErr            error
+type AWSMetadataAPIStub struct {
```
Comment (daniel-weisse) [inline internal/attestation/aws/issuer_test.go:106]:
```suggestion
type stubMetadataAPI struct {
```
Struct should not be exported.
For testing types we usually use the `stub<Type>` naming scheme

File: internal/attestation/aws/validator_test.go [hunk 10]
```diff
@@ -7,13 +7,17 @@ SPDX-License-Identifier: AGPL-3.0-only
 package aws
 
 import (
+	"context"
 	"encoding/json"
+	"errors"
 	"testing"
-	"time"
+
+	//"github.com/aws/aws-sdk-go-v2/config"
```
Comment (daniel-weisse) [inline internal/attestation/aws/validator_test.go:15]:
```suggestion
```

File: internal/attestation/aws/validator_test.go [hunk 11]
```diff
@@ -114,65 +118,21 @@ func TestTpmEnabled(t *testing.T) {
...
-	mockDocTpm.Region = "us-east-2"
-
-	mockDocNoTpm := mockDocBase
-	mockDocNoTpm.ImageID = "ami-0b59bfac6be064b78"
-	mockDocTpm.Region = "us-east-2"
-
-	attDocTpm := vtpm.AttestationDocument{}
-	attDocTpm.UserData, _ = json.Marshal(mockDocTpm)
-	attDocTpm.Attestation = &attest.Attestation{}
+type AWSMetadataStub struct {
```
Comment (daniel-weisse) [inline internal/attestation/aws/validator_test.go:121]:
```suggestion
type stubDescribeAPI struct {
```
Same here.

File: internal/attestation/aws/validator_test.go [hunk 12]
```diff
@@ -114,65 +118,21 @@ func TestTpmEnabled(t *testing.T) {
...
+	describeImagesErr        error
+	describeImagesTPMSupport string
+}
 
-	attDocNoTpm := vtpm.AttestationDocument{}
-	attDocNoTpm.UserData, _ = json.Marshal(mockDocNoTpm)
-	attDocNoTpm.Attestation = &attest.Attestation{}
+// DescribeImages is a mock function for testing.
+// Although the original function works for multiple supplied image id's, this function ignores this, since in our code we also only get the information for one image.
```
Comment (daniel-weisse) [inline internal/attestation/aws/validator_test.go:127]: Methods on test types don't need their own doc comment, unless they do something overly complex which should be documentet, or the type is exported for use in other modules.
Other wise looks good ;)

File: internal/attestation/aws/validator.go [hunk 13]
```diff
@@ -21,23 +20,31 @@ import (
...
+		enforcedPCRs,
+		getTrustedKey,
+		v.tpmEnabled,
+		vtpm.VerifyPKCS1v15,
+		log,
+	)
+	client, err := config.LoadDefaultConfig(context.Background(), config.WithEC2IMDSRegion())
+	if err != nil {
+		return nil
 	}
```
Comment (daniel-weisse) [inline internal/attestation/aws/validator.go:44]: I didn't consider the implications for existing code if we create the client in the Validator init function...
In a lot of places our code assumes the `NewValidator` functions cannot return errors and the returned value is always valid (e.g. not nil).

Made a suggestion PR to fix this, please check it

**Review by daniel-weisse — COMMENTED** [review]
Looks good overall
There are some linting issues which should be reported by the CI once merge conflicts are resolved

**Comment by daniel-weisse** [comment]
Rebase/fix merge conflicts and we can merge this to main :whale:

**Review by daniel-weisse — APPROVED** [review]
